src/market/directory.rs

```diff
@@ -191,64 +185,83 @@ pub async fn start_directory_server(directory: Arc<DirectoryServer>) {
-                }
-            }
-            _ = tokio::time::sleep(Duration::from_secs(3)) => {
-                if *directory.shutdown.read().unwrap() {
-                    log::info!("Shutdown signal received. Stopping directory server.");
-                    if directory.connection_type == ConnectionType::TOR && cfg!(feature = "tor"){
-                        crate::tor::kill_tor_handles(handle.unwrap());
-                        log::info!("Directory server and Tor instance terminated successfully");
-                    }
-                    break;
-                } else {
-                     let file_content = addresses.read().unwrap().iter().map(|addr| {
-                        format!("{}\n", addr)
-                    }).collect::<Vec<String>>().join("");
-                    let mut file = OpenOptions::new()
-                    .write(true)
-                    .create(true)
-                    .append(true)
-                    .open(address_file.to_str().unwrap())
-                    .await
+    let addres_arc = addresses.clone();
+    let rpc_thread = thread::spawn(|| {
+        start_rpc_server_thread(directory_server_arc, addres_arc);
+    });
+
+    let listener = TcpListener::bind((Ipv4Addr::LOCALHOST, directory.port)).unwrap();
+
+    while !*directory.shutdown.read().unwrap() {
+        match listener.accept() {
+            Ok((mut stream, addrs)) => {
+                log::debug!("Incoming connection from : {}", addrs);
+                let address_arc = addresses.clone();
+                stream
+                    .set_read_timeout(Some(Duration::from_secs(20)))
                     .unwrap();
-                    file.write_all(file_content.as_bytes()).await.unwrap();
-                }
+                stream
+                    .set_write_timeout(Some(Duration::from_secs(20)))
+                    .unwrap();
+                handle_client(&mut stream, address_arc);
+            }
+
+            // If no connection received, check for shutdown or save addresses to disk
+            Err(e) => {
+                log::error!("Error accepting incoming connection: {:?}", e);
             }
         }
+
+        sleep(Duration::from_secs(3));
     }
+
+    log::info!("Shutdown signal received. Stopping directory server.");
+    rpc_thread.join().unwrap();
+    if let Some(handle) = tor_handle {
+        crate::tor::kill_tor_handles(handle);
+        log::info!("Directory server and Tor instance terminated successfully");
+    }
+
+    // Write the addresses to file
+    let file_content = addresses
+        .read()
+        .unwrap()
+        .iter()
+        .map(|addr| format!("{}\n", addr))
+        .collect::<Vec<String>>()
+        .join("");
+    let mut file = OpenOptions::new()
+        .write(true)
+        .create(true)
+        .truncate(true) // Override the address file
+        .open(address_file.to_str().unwrap())
+        .unwrap();
+    file.write_all(file_content.as_bytes()).unwrap();
```

Comment:
Yes. Nice pointer. We should periodically save the data on the disk to avoid loss. Although that would need its own thread. Very simply one though, another `while !*directory.shutdown.read()`.

This should be done in a follow-up PR to not clutter the change set here too much.